llvm/lib/Target/AMDGPU/AMDGPUIGroupLP.cpp

```diff
@@ -904,6 +909,854 @@ void MFMASmallGemmOpt::applyIGLPStrategy(
+          ++TransPipeCount;
+          break;
+        }
+      }
+    }
+
+    if (!TempExp.has_value())
+      return;
+
+    // Count the number of MFMAs that are reached by an EXP
+    for (auto &SuccSU : MFMAPipeCands) {
+      if (std::find_if(MFMAPipeSUs.begin(), MFMAPipeSUs.end(),
+                       [&SuccSU](SUnit *PotentialMatch) {
+                         return PotentialMatch == SuccSU;
+                       })) {
+        ++MFMAPipeCount;
+        continue;
+      }
+      for (auto &PredSU : ExpPipeCands) {
+        if (DAG->IsReachable(SuccSU, PredSU)) {
+          ++MFMAPipeCount;
+          break;
+        }
+      }
+    }
+
+    if (!TempMFMA.has_value() || !TempExp.has_value())
+      return;
+
+    // The number of bit pack operations that depend on a single V_EXP
+    unsigned PackSuccCount = std::count_if(
+        PackSUs.begin(), PackSUs.end(), [this, &TempExp](SUnit *VPack) {
+          return DAG->IsReachable(VPack, *TempExp);
+        });
+
+    // The number of bit pack operations an MFMA depends on
+    unsigned PackPredCount =
+        std::count_if((*TempMFMA)->Preds.begin(), (*TempMFMA)->Preds.end(),
+                      [&isBitPack](SDep &Pred) {
+                        auto Opc = Pred.getSUnit()->getInstr()->getOpcode();
+                        return isBitPack(Opc);
+                      });
+
+    auto PackPred =
+        std::find_if((*TempMFMA)->Preds.begin(), (*TempMFMA)->Preds.end(),
+                     [&isBitPack](SDep &Pred) {
+                       auto Opc = Pred.getSUnit()->getInstr()->getOpcode();
+                       return isBitPack(Opc);
+                     });
+
+    // How many MFMAs depend on a single bit pack operation
+    MFMAEnablement =
+        std::count_if(PackPred->getSUnit()->Succs.begin(),
+                      PackPred->getSUnit()->Succs.end(), [&TII](SDep &Succ) {
+                        return TII->isMFMAorWMMA(*Succ.getSUnit()->getInstr());
+                      });
+
+    // The number of MFMAs that depend on a single V_EXP
+    MFMAEnablement *= PackSuccCount;
+
+    // The number of V_EXPs required to resolve all dependencies for an MFMA
+    ExpRequirement =
+        std::count_if(ExpPipeCands.begin(), ExpPipeCands.end(),
+                      [this, &PackPred](SUnit *ExpBase) {
+                        return DAG->IsReachable(PackPred->getSUnit(), ExpBase);
+                      });
+
+    ExpRequirement *= PackPredCount;
+  }
+
+  bool IsSmallKernelType =
+      MFMAEnablement == 2 && ExpRequirement == 4 && TransPipeCount == 32;
+  bool IsLargeKernelType =
+      MFMAEnablement == 4 && ExpRequirement == 4 && TransPipeCount == 64;
+  bool IsTinyKernelType =
+      MFMAEnablement == 1 && ExpRequirement == 4 && TransPipeCount == 32;
+
+  if (!(IsSmallKernelType || IsLargeKernelType || IsTinyKernelType))
+    return;
+
+  unsigned PipelineSyncID = 0;
+  SchedGroup *SG = nullptr;
+
+  if (IsSmallKernelType && Phase != IGLPPhase::PostRA) {
```

Comment:
This function is way too long and repetitive. Do we really need so many instances of these rules?

Reply:
I'll see if anything can be done regarding the amount of SchedGroups / rules. 

We do plan to replace the implementation with a cleaner generalized version which will address this concern. That said, we want to offer an incremental version to capture some performance.

Reply:
I have a version of the generalized implementation which resolves this issue (too long and repetitive), while capturing the same performance. I plan to push that as a separate PR to compete with this one.

Reply:
See https://github.com/llvm/llvm-project/pull/81342